wallet/src/Dscp/Wallet/Knit.hs

```diff
@@ -0,0 +1,170 @@
+module Dscp.Wallet.Knit where
+
+import Prelude hiding (preview)
+
+import Codec.Serialise (deserialiseOrFail)
+import Control.Exception
+import Control.Lens
+import Dscp.Core (addrFromText, txId)
+import Dscp.Crypto (mkPassPhrase)
+import IiExtras
+import Text.Earley
+
+import qualified Data.ByteString.Lazy as BSL
+import qualified Serokell.Util.Base64 as Base64
+import qualified Text.Megaparsec as P
+import qualified Text.Megaparsec.Char as P
+import qualified Text.Megaparsec.Char.Lexer as P
+
+import Dscp.Wallet.Face
+
+import Knit
+
+-- Component type for Knit
+data Wallet
+
+data instance ComponentValue _ Wallet
+  = ValueAddress Address
+  | ValueCoin Coin
+  | ValueSecretKey ByteString
+  deriving (Eq, Ord, Show)
+
+makePrisms 'ValueAddress
+
+instance Elem components Wallet => ComponentInflate components Wallet where
+  componentInflate = \case
+    ValueAddress a -> ExprLit $ toLit (LitAddress a)
+    ValueCoin c -> ExprLit $ toLit (LitCoin c)
+    ValueSecretKey sk -> ExprLit $ toLit (LitSecretKey sk)
+
+data instance ComponentLit Wallet
+  = LitAddress Address
+  | LitCoin Coin
+  | LitSecretKey ByteString
+  deriving (Eq, Ord, Show)
+
+data instance ComponentToken Wallet
+  = TokenAddress Address
+  | TokenCoin Coin
+  | TokenSecretKey ByteString
+  deriving (Eq, Ord, Show)
+
+makePrisms 'TokenAddress
+
+instance Elem components Wallet => ComponentTokenizer components Wallet where
+  componentTokenizer =
+      [ toToken . TokenAddress <$> pAddress
+      , toToken . TokenCoin <$> pCoin
+      , toToken . TokenSecretKey <$> pSecretKey
+      ]
+    where
+      pAddress :: Tokenizer Address
+      pAddress = do
+        str <- pSomeAlphaNum
+        either (fail . show) return $ addrFromText str
+
+      pCoin :: Tokenizer Coin
+      pCoin = Coin <$> P.signed (return ()) P.decimal
+
+      pSecretKey :: Tokenizer ByteString
+      pSecretKey = do
+        str <- (<>) <$> P.takeP (Just "base64") 91 <*> P.string "="
```

Comment:
Why can't you read till e.g. a non-letter/EOF?
If that's needed for performance, please move the constant next to `SecretKey` type.

Reply:
Btw encoded secret key takes 44 chars for me with the trailing `=`, how could it be :thinking:

Reply:
Unfortunately, it will consider _any_ base64 string as secret key then.
Knit is still somewhat weirdly typed language.

For Cardano we "solved" it by requiring `#` symbol before it.
I think in this case I'll just require all keys and addresses to require quotes around them as normal strings, and leave all validation to command representation below, like with passphrases.

Reply:
I generated secret key using `keyGen`, `mkPassPhrase` and `encrypt`, got a string like this:
`UH1OoUs/eDxxKHMlLy09+m5Qw67i+iDVvJ6wZora1TDY6FgggZoDtW8FW7TD0iUx99vPg4q4mUydmMRZLiT9bA+sEug=`

Reply:
Actually, disregard that, I can do it the same way as addresses.

Reply:
> I generated secret key using keyGen, mkPassPhrase and encrypt, got a string like this:
UH1OoUs/eDxxKHMlLy09+m5Qw67i+iDVvJ6wZora1TDY6FgggZoDtW8FW7TD0iUx99vPg4q4mUydmMRZLiT9bA+sEug=
Right, I forgot to encrypt secret key :man_facepalming: